### PR TITLE
SNOW-1445867: Add support for `DataFrame.pivot` and `pd.pivot`

### PR DESCRIPTION
**Pull request by sfc-gh-nkrishna**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1445867

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This PR adds support for `DataFrame.pivot` and `pd.pivot` leveraging the `DataFrame.pivot_table` implementation.

### CODE REVIEW

File: src/snowflake/snowpark/modin/pandas/dataframe.py [hunk 1]
```diff
@@ -1813,12 +1813,31 @@ def unstack(self, level=-1, fill_value=None):  # noqa: PR01, RT01, D200
...
+
+        # if values is not specified, it should be the remaining columns not in
+        # index or columns
+        if values is None:
+            values = list(self.columns)
+            if index is not None:
+                values = [v for v in values if v not in index]
+            if columns is not None:
+                values = [v for v in values if v not in columns]
+
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/pandas/dataframe.py:1840]: Is this code copied from latest modin?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -6804,6 +6804,48 @@ def is_series_like(self) -> bool:
...
+            Column(s) to use for populating new frame’s values. If not specified, all remaining columns
+            will be used and the result will have hierarchically indexed columns.
+
+        Returns
+        -------
+        SnowflakeQueryCompiler
+        """
+        # Call pivot_table which is a more generalized version of pivot with `min` aggregation
+        # Note we differ from pandas by not checking for duplicates and raising a ValueError as that would require an eager query
+        return self.pivot_table(
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:6836]: For error scenarios like invalid labels for columns, index etc, are errors raised from `pivot_table` same as `pivot`?

File: src/snowflake/snowpark/modin/pandas/dataframe.py (hunk 1 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/pandas/dataframe.py:1840]: Yes this is copied straight from latest Modin: https://github.com/modin-project/modin/blob/main/modin/pandas/dataframe.py

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 2 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:6836]: Yes the error scenarios are the exact same, `pivot_table` is a generalized version of `pivot`. The only difference is what is described in the comment around checking for duplicates/pre-empting with a ValueError. In other APIs we don't do similar checks due to the same lazy execution rationale.

**Review by sfc-gh-nkumar — APPROVED** [review]
LGTM, thanks!

File: src/snowflake/snowpark/modin/pandas/dataframe.py [hunk 3]
```diff
@@ -1813,12 +1813,31 @@ def unstack(self, level=-1, fill_value=None):  # noqa: PR01, RT01, D200
...
             )
 
-    @dataframe_not_implemented()
-    def pivot(self, index=None, columns=None, values=None):  # noqa: PR01, RT01, D200
+    def pivot(
+        self,
+        *,
+        columns: Any,
+        index: Any | NoDefault = no_default,
+        values: Any | NoDefault = no_default,
```
Comment (sfc-gh-jkew) [inline src/snowflake/snowpark/modin/pandas/dataframe.py:1821]: Does values have to be hashable? Or are we pretty sure it is hashable already?

**Review by sfc-gh-jkew — APPROVED** [review]
LGTM but I was wondering what would happen if values is not hashable.

File: src/snowflake/snowpark/modin/pandas/dataframe.py (hunk 3 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/pandas/dataframe.py:1821]: On the pandas docs, they describe values as `str, object or a list of the previous`. So in theory you could have an object passed that is not hashable.

In `pivot_table`, we currently check whether Index, Columns, and Values are hashable and if not raise a NotImplementedError. Currently, we don't support anything other than strings or list of strings at the moment but have this TODO marked in the code: `TODO: SNOW-857485 Support for non-str and list of non-str for index/columns/values`

**Pull request merged**